edk2toolext/environment/reporttypes/coverage_report.py

```diff
@@ -0,0 +1,339 @@
+        group.add_argument("--by-package", action="store_true", dest="by_package", default=False,
+                           help="Filters test coverage to only files in the specified packages(s)")
+        group.add_argument("-p", "--package", "--Package", "--PACKAGE", dest="package_list",
+                           action=SplitCommaAction, default=[],
+                           help="The package to include in the report. Can be specified multiple times.")
+        # Group 2
+        group = parserobj.add_argument_group("Coverage by platform options")
+        group.add_argument("--by-platform", action="store_true", dest="by_platform", default=False,
+                           help="Filters test coverage to all files used to build the specified platform package.")
+        group.add_argument("-d", "--dsc", "--DSC", dest="dsc",
+                           help="Edk2 relative path the ACTIVE_PLATFORM DSC file.")
+
+        # Other args
+        parserobj.add_argument(dest="xml", action="store", help="The path to the XML file parse.")
+        parserobj.add_argument("-o", "--output", "--Output", "--OUTPUT", dest="output", default="Coverage.xml",
+                               help="The path to the output XML file.", action="store")
+        parserobj.add_argument("-ws", "--workspace", "--Workspace", "--WORKSPACE", dest="workspace",
+                               help="The Workspace root associated with the xml argument.", default=".")
+
+        parserobj.add_argument("-e", "--exclude", "--Exclude", "--EXCLUDE", dest="exclude",
+                               action=SplitCommaAction, default=[],
+                               help="Package path relative paths or file. Globbing is supported. Can be specified "
+                               "multiple times")
+
+    def run_report(self, db: Edk2DB, args: Namespace) -> None:
+        """Generate the Coverage report."""
+        self.args = args
+        if args.by_package:
+            logging.info("Organizing coverage report by Package.")
+            return self.run_by_package(db)
+        if args.by_platform:
+            logging.info("Organizing coverage report by Platform.")
+            return self.run_by_platform(db)
+
+        logging.error("No report type specified via command line or configuration file.")
+        return -1
+
+    def run_by_platform(self, db: Edk2DB):
+        """Runs the report, only adding coverage data for source files used to build the platform.
+
+        Args:
+            db (Edk2DB): The database containing the necessary data
+
+        Returns:
+            (bool): True if the report was successful, False otherwise.
+        """
+        # Verify valid ACTIVE_PLATFORM
+        dsc = self.args.dsc
+        if not dsc:
+            logging.error("No ACTIVE_PLATFORM dsc file specified. It should be edk2 package path relative.")
+            return -1
+        logging.info(f"ACTIVE_PLATFORM requested: {dsc}")
+
+        # replace any files with their contents
```

Comment:
This is interpreting the exclusion list as pointing to a file, but for the package, we dont do this. Is this intentionally different?

Reply:
Good catch. It was accidently left out of run_by_package. I'll move it up to run_report, so that both format types (package / platform) can consume the exclude list.

I'll also note that it supports a mixture of directly defining the pattern in the cli, providing a file of patterns, and a mixture.

i.e. `-e pattern1 -e file_with_patterns.txt -e pattern2` or `-e pattern1,file_with_patterns.txt,pattern2`

Reply:
Created a function and moved its call to `run_report`. Thanks!